File: python/event.py
'''Import Txn, Blk, Peer and Simulator'''
from txn import Txn
from block import Blk
from peer import Peer
from simulator import Simulator

class Event:
    '''Event'''

    # construct event
    def __init__(self, timestamp, typ, peer:Peer, txn:Txn, fro:Peer, blk:Blk):
        '''constructor'''
        self.timestamp = timestamp
        self.type = typ
        self.peer = peer
        self.txn = txn
        self.fro = fro
        self.blk = blk

    # runner
    def run(self, sim:Simulator):
        '''event runner'''
        if self.type == 1:
            self.peer.generate_txn(sim, self)
        elif self.type == 2:
            self.peer.forward_txn(sim, self)
        elif self.type == 3:
            self.peer.hear_txn(sim, self)
        elif self.type == 4:
            self.peer.generate_blk(sim, self)
        elif self.type == 5:
            self.peer.forward_blk(sim, self)
        elif self.type == 6:
            self.peer.hear_blk(sim, self)
        elif self.type == 7:
            self.peer.update_tree(sim, self)

File: assgn1/200020120_200050129_200050157/code/event.cpp
#include "include/header.hpp"

/**
 * @brief Constructor for the Event Class
 *
 * @param timestamp Timestamp of the event
 * @param type Type of event (Code given from 1 to 7)
 * @param p Peer associated with the event (may be generator, sender or reciever)
 * @param tran Transaction associated with the event
 * @param from Peer who sent the event (iff applicable)
 * @param block Block Associated with the event
 */
event::event(ld timestamp, int type, peer* p/*=nullptr*/, txn* tran/*=nullptr*/, peer* from/*=nullptr*/, blk* block/*=nullptr*/) {
    this->timestamp = timestamp;
    this->type = type;
    this->p = p;
    this->tran = tran;
    this->from = from;
    this->block = block;
}

/**
 * @brief Function to run the event
 * Checks the type of event it is and run the appropriate functions 
 *
 * @param sim Simulation which has the global variables
 */
void event::run(simulator& sim) {
	
    cout << "event run: type=" << type << '\n';
    switch (type) {
        case 1: {
            p->generate_txn(sim, this);
            break;
        }
        case 2: {
            // if(tran==nullptr) exit(1);
            // cout << this->tran->txn_id << '\n';
            p->forward_txn(sim, this);
            break;
        }
        case 3: {
            // if(from==nullptr || tran==nullptr) exit(1);
            // cout << this->tran->txn_id << '\n';
            p->hear_txn(sim, this);
            break;
        }
        case 4: {
            p->generate_blk(sim, this);
            break;
        }
        case 5: {
            p->forward_blk(sim, this);
            break;
        }
        case 6: {
            p->hear_blk(sim, this);
            break;
        }
        case 7: {
            p->update_tree(sim, this);
            break;
        }
        default: {
            cout << "incorrect event type: " << type << '\n';
            break;
        }
    }
}
